### PR TITLE
[TierTwo] Introduce network requests manager

### PR DESCRIPTION
**Pull request by furszy**
Built on top of #2684

An object in charge of cache and store tier two network requests information, preventing possible bans for asking the same data too often.

### CODE REVIEW

**Comment by furszy** [comment]
Rebased, conflicts and compilation issue solved. Ready to go.

**Review by random-zebra — COMMENTED** [review]
tested ACK.
Aside a couple of nits, it seems that the netfulfilled manager is not persisted to disk. We should probably add
```
CFlatDB<CNetFulfilledRequestManager>("mnetrequests.dat", "magicNetRequestsCache").Dump(g_netfulfilledman);
```
to `DumpTierTwo()`.

**Comment by furszy** [comment]
good catch, adding the dump.

**Comment by furszy** [comment]
Updated per feedback.

File: src/tiertwo/init.cpp [hunk 1]
```diff
@@ -133,6 +133,23 @@ bool LoadTierTwo(int chain_active_height, bool fReindexChainState)
...
+    if (load_cache_files) {
+        if (!netRequestsDb.Load(g_netfulfilledman)) {
+            LogPrintf("Failed to load network requests cache from %s", (pathDB / strDBName).string());
+        }
+    } else {
+        CNetFulfilledRequestManager netfulfilledmanTmp;
+        if (!netRequestsDb.Dump(netfulfilledmanTmp)) {
+            LogPrintf("Failed to clear network requests cache at %s", (pathDB / strDBName).string());
+        }
+    }
```
Comment (random-zebra) [inline src/tiertwo/init.cpp:151]:
```patch
Index: src/tiertwo/init.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/tiertwo/init.cpp b/src/tiertwo/init.cpp
--- a/src/tiertwo/init.cpp	(revision 4b91b0cdc61269e434104b247dc75ebd8873b39a)
+++ b/src/tiertwo/init.cpp	(date 1644481131458)
@@ -136,17 +136,16 @@
     // ############################## //
     // ## Network Requests Manager ## //
     // ############################## //
-    std::string strDBName = "netrequests.dat";
     uiInterface.InitMessage(_("Loading network requests cache..."));
-    CFlatDB<CNetFulfilledRequestManager> netRequestsDb(strDBName, "magicNetRequestsCache");
+    CFlatDB<CNetFulfilledRequestManager> netRequestsDb(NET_REQUESTS_CACHE_FILENAME, NET_REQUESTS_CACHE_FILE_ID);
     if (load_cache_files) {
         if (!netRequestsDb.Load(g_netfulfilledman)) {
-            LogPrintf("Failed to load network requests cache from %s", (pathDB / strDBName).string());
+            LogPrintf("Failed to load network requests cache from %s", netRequestsDb.GetDbPath().string());
         }
     } else {
         CNetFulfilledRequestManager netfulfilledmanTmp;
         if (!netRequestsDb.Dump(netfulfilledmanTmp)) {
-            LogPrintf("Failed to clear network requests cache at %s", (pathDB / strDBName).string());
+            LogPrintf("Failed to clear network requests cache at %s", netRequestsDb.GetDbPath().string());
         }
     }
```

File: src/tiertwo/netfulfilledman.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#define PIVX_NETFULFILLEDMAN_H
+
+#include "serialize.h"
+#include "sync.h"
+
+#include <map>
+
+class CService;
+
+static const std::string NET_REQUESTS_CACHE_FILENAME = "mnetrequests.dat";
```
Comment (random-zebra) [inline src/tiertwo/netfulfilledman.h:16]:
```suggestion
static const std::string NET_REQUESTS_CACHE_FILENAME = "netrequests.dat";
```
according to `strDBName` in `LoadTierTwo`

**Review by random-zebra — COMMENTED** [review]
With the suggested change the argument `pathDB` of `LoadTierTwo` can be removed.

**Comment by furszy** [comment]
great zebra 👍 , suggestions applied.

**Review by random-zebra — APPROVED** [review]
ACK be1c0fcc7605ba4e668433cda2212c97654bce57

**Review by Fuzzbawls — APPROVED** [review]
ACK be1c0fcc7605ba4e668433cda2212c97654bce57

**Pull request merged**